File: core/src/modules/console_module.cc
/*
 *
 * Tencent is pleased to support the open source community by making
 * Hippy available.
 *
 * Copyright (C) 2019 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "core/modules/console_module.h"

#include <string>

#include "base/logging.h"
#include "core/base/string_view_utils.h"
#include "core/scope.h"


using unicode_string_view = tdf::base::unicode_string_view;
using Ctx = hippy::napi::Ctx;
using CtxValue = hippy::napi::CtxValue;
using StringViewUtils = hippy::base::StringViewUtils;

GEN_INVOKE_CB(ConsoleModule, Log) // NOLINT(cert-err58-cpp)

namespace {

unicode_string_view EscapeMessage(const unicode_string_view& str_view) {
  unicode_string_view::u8string u8_str = StringViewUtils::Convert(str_view, unicode_string_view::Encoding::Utf8).utf8_value();
  size_t len = u8_str.length();
  unicode_string_view::u8string ret;
  for (size_t i = 0; i < len; i++) {
    auto c = u8_str[i];
    ret += c;
    if (c == '%') {
      ret += '%';
    }
  }

  return unicode_string_view(ret);
}

}  // namespace

void ConsoleModule::Log(const hippy::napi::CallbackInfo& info, void* data) { // NOLINT(readability-convert-member-functions-to-static)
  auto scope_wrapper = reinterpret_cast<ScopeWrapper*>(std::any_cast<void*>(info.GetSlot()));
  auto scope = scope_wrapper->scope.lock();
  TDF_BASE_CHECK(scope);
  auto context = scope->GetContext();
  TDF_BASE_CHECK(context);

  unicode_string_view message;
  if (!context->GetValueString(info[0], &message)) {
    info.GetExceptionValue()->Set(context,"The first argument must be string.");
    return;
  }

  unicode_string_view view_msg = EscapeMessage(message);
  if (info.Length() == 1) {
    TDF_BASE_LOG(INFO) << view_msg;
  } else {
    unicode_string_view view_type;
    if (!context->GetValueString(info[1], &view_type) ||
        StringViewUtils::IsEmpty(view_type)) {
      info.GetExceptionValue()->Set(
          context, "The second argument must be non-empty string.");
      return;
    }

    std::string u8_type = StringViewUtils::ToU8StdStr(view_type);
    if (u8_type == "info") {
      TDF_BASE_LOG(INFO) << view_msg;
    } else if (u8_type == "warn") {
      TDF_BASE_LOG(WARNING) << view_msg;
    } else if (u8_type == "error") {
      TDF_BASE_LOG(ERROR) << view_msg;
    } else if (u8_type == "fatal") {
      TDF_BASE_LOG(FATAL) << view_msg;
    } else {
      TDF_BASE_LOG(INFO) << view_msg;
    }
  }

  info.GetReturnValue()->SetUndefined();
}

std::shared_ptr<CtxValue> ConsoleModule::BindFunction(std::shared_ptr<Scope> scope,
                                                      std::shared_ptr<CtxValue>* rest_args) {
  auto context = scope->GetContext();
  auto object = context->CreateObject();

  auto key = context->CreateString("Log");
  auto wrapper = std::make_unique<hippy::napi::FuncWrapper>(
      InvokeConsoleModuleLog,nullptr);
  auto value = context->CreateFunction(wrapper);
  scope->SaveFuncWrapper(std::move(wrapper));
  context->SetProperty(object, key, value);

  return object;
}

File: core/js/global/ConsoleModule.js
/*
 * Tencent is pleased to support the open source community by making
 * Hippy available.
 *
 * Copyright (C) 2017-2022 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable no-undef */
/* eslint-disable no-underscore-dangle */

const consoleModule = internalBinding('ConsoleModule');

const indent = (level) => {
  let tab = '';
  while (tab.length < level * 2) {
    tab += '  ';
  }
  return tab;
};

let inspectObject = null;
let inspect = null;

function getOwnPropertyDescriptors(param) {
  const result = {};
  const propKeys = Object.keys(param);
  propKeys.forEach((key, index) => {
    result[propKeys[index]] = Object.getOwnPropertyDescriptor(param, propKeys[index]);
  });
  return result;
}

inspectObject = (value, level = 0, recurseTimes = 2, linebreak = '\n') => {
  if (value === null) {
    return 'null';
  }
  if (Array.isArray(value)) {
    return `[ ${value.map(item => inspect(item, level + 1, recurseTimes - 1)).join(', ')} ]`;
  }

  switch (Object.prototype.toString.call(value)) {
    case '[object Date]': {
      return Date.prototype.toISOString.call(value);
    }
    case '[object RegExp]': {
      return RegExp.prototype.toString.call(value);
    }
    case '[object Error]': {
      return `${value.stack || Error.prototype.toString.call(value)}`;
    }
    case '[object Set]': {
      return `Set { ${[...value].map(item => inspect(item, level + 1, recurseTimes - 1, linebreak)).join(', ')} }`;
    }
    case '[object WeakSet]': {
      return 'WeakSet { [items unknown] }';
    }
    case '[object Map]': {
      const entries = [];
      value.forEach((val, key) => {
        entries.push(`${inspect(key, recurseTimes * -1, recurseTimes - 1, '')} => ${inspect(val, recurseTimes * -1, recurseTimes - 1, '')}`);
      });
      return `Map { ${entries.join(', ')} }`;
    }
    case '[object WeakMap]': {
      return 'WeakMap { [items unknown] }';
    }
    default: {
      break;
    }
  }

  if (value instanceof Error) {
    return `${value.stack || Error.prototype.toString.call(value)}`;
  }
  const descs = getOwnPropertyDescriptors(value);
  const keys = Object.keys(descs);
  const pairs = [];
  keys.forEach((key) => {
    const desc = descs[key];
    const itemPrefix = `${indent(level + 1)}${key} : `;
    if (desc.get) {
      if (desc.set) {
        pairs.push(`${itemPrefix}[Getter/Setter]`);
      } else {
        pairs.push(`${itemPrefix}[Getter]`);
      }
    } else if (desc.set) {
      pairs.push(`${itemPrefix}[Setter]`);
    } else if (desc.value) {
      pairs.push(`${itemPrefix}${inspect(desc.value, level + 1, recurseTimes - 1, linebreak)}`);
    }
  });
  return `{${linebreak}${pairs.join(`, ${linebreak}`)}${linebreak}${indent(level)}}`;
};

inspect = (value, level = 0, recurseTimes = 2, linebreak = '\n') => {
  switch (typeof value) {
    case 'string':
      return `'${value}'`;
    case 'symbol':
      return value.toString();
    case 'function':
      return `[Function${value.name ? `: ${value.name}` : ''}]`;
    case 'object':
      if (recurseTimes < 0) {
        return '[Object]';
      }
      return inspectObject(value, level, recurseTimes, linebreak);
    case 'bigint':
      return `n${value}`;
    case 'undefined':
    case 'number':
    case 'boolean':
    default:
      return `${value}`;
  }
};

let vmConsole;
if (typeof console !== 'undefined') {
  vmConsole = console;
}

const supportApiList = ['log', 'info', 'warn', 'error', 'debug'];
global.ConsoleModule = {};
supportApiList.forEach((api) => {
  global.ConsoleModule[api] = (...args) => {
    const log = args.map(arg => inspect(arg)).join(' ');
    consoleModule.Log(log, api);
  };
});

global.console = {
  reportUncaughtException(error) {
    if (error && error instanceof Error) {
      throw error;
    }
  },
};

if (vmConsole) {
  Object.keys(vmConsole).forEach((api) => {
    global.console[api] = vmConsole[api];
  });
}
